File: sorting/bubble_sort/bubble_sort_debug.rb
# Bubble Sort Implementation in Ruby with Debug Points
# Time Complexity: O(n²)
# Space Complexity: O(1)

def bubble_sort(arr)
  n = arr.length
  swapped = true
  iteration = 0
  
  puts "🔍 DEBUG: Starting bubble sort with array: #{arr.inspect}"
  
  while swapped
    iteration += 1
    swapped = false
    
    puts "🔄 DEBUG: Iteration #{iteration}, array: #{arr.inspect}"
    
    (0...n-1).each do |i|
      puts "  📊 DEBUG: Comparing arr[#{i}]=#{arr[i]} with arr[#{i+1}]=#{arr[i+1]}"
      
      if arr[i] > arr[i + 1]
        puts "  🔄 DEBUG: Swapping #{arr[i]} and #{arr[i+1]}"
        # Swap elements
        arr[i], arr[i + 1] = arr[i + 1], arr[i]
        swapped = true
        puts "  ✅ DEBUG: After swap: #{arr.inspect}"
      else
        puts "  ✅ DEBUG: No swap needed"
      end
    end
    
    n -= 1  # Last element is now in place
    puts "🎯 DEBUG: End of iteration #{iteration}, swapped: #{swapped}, n: #{n}"
  end
  
  puts "🎉 DEBUG: Sorting completed in #{iteration} iterations"
  arr
end

# Example usage with debug output
if __FILE__ == $0
  test_array = [64, 34, 25, 12, 22, 11, 90]
  puts "🚀 Original array: #{test_array}"
  
  sorted_array = bubble_sort(test_array.dup)
  puts "✅ Final sorted array: #{sorted_array}"
end

File: sorting/bubble_sort/bubble_sort_debug.ts
// Bubble Sort Implementation in TypeScript with Debug Points
// Time Complexity: O(n²)
// Space Complexity: O(1)

function bubbleSort(arr: number[]): number[] {
    const n = arr.length;
    let swapped = true;
    let iteration = 0;
    
    console.log(`🔍 DEBUG: Starting bubble sort with array: [${arr}]`);
    
    while (swapped) {
        iteration += 1;
        swapped = false;
        
        console.log(`🔄 DEBUG: Iteration ${iteration}, array: [${arr}]`);
        
        for (let i = 0; i < n - 1; i++) {
            console.log(`  📊 DEBUG: Comparing arr[${i}]=${arr[i]} with arr[${i+1}]=${arr[i+1]}`);
            
            if (arr[i] > arr[i + 1]) {
                console.log(`  🔄 DEBUG: Swapping ${arr[i]} and ${arr[i+1]}`);
                // Swap elements
                [arr[i], arr[i + 1]] = [arr[i + 1], arr[i]];
                swapped = true;
                console.log(`  ✅ DEBUG: After swap: [${arr}]`);
            } else {
                console.log(`  ✅ DEBUG: No swap needed`);
            }
        }
        
        n--;  // Last element is now in place
        console.log(`🎯 DEBUG: End of iteration ${iteration}, swapped: ${swapped}, n: ${n}`);
    }
    
    console.log(`🎉 DEBUG: Sorting completed in ${iteration} iterations`);
    return arr;
}

// Example usage with debug output
if (require.main === module) {
    const testArray: number[] = [64, 34, 25, 12, 22, 11, 90];
    console.log(`🚀 Original array: [${testArray}]`);
    
    const sortedArray = bubbleSort([...testArray]);
    console.log(`✅ Final sorted array: [${sortedArray}]`);
}

export default bubbleSort;
